[ZT] Access for SaaS OIDC (#20603)
cloudflare/cloudflare-docs: src/content/docs/cloudflare-one/applications/configure-apps/saas-apps/generic-oidc-saas.mdx

   | Scope     | Description                                          |
   | --------- | ---------------------------------------------------- |
   | `openid`  | Includes a unique identifier for the user (required). |
   | `email`   | Includes the user's email address.                    |
   | `profile` | Includes the user's name and all custom OIDC claims from the IdP.         |
   | `groups`  | Include the user's IdP group membership.             |

In your Access application, you can configure the OIDC scopes and claims that Access sends to the SaaS provider. For example, you can remove the `groups` scope if your SaaS application does not need to receive user group information.

### Filter groups

In **Group filter regex**, you can enter a regular expression to define the identity provider groups that you want to include in the `groups` scope. For example, if you enter the expression `(^TEAM-Engineering-.$)|(^TEAM-Product-.$)`, only groups with names like  TEAM-Engineering-A or TEAM-Product-B would get passed to the SaaS application.

### Add claims

> To add additional OIDC claims to your SaaS application, configure the following fields for each claim:

Reviewer: ```suggestion
To add additional OIDC claims onto the ID Token sent to your SaaS application, configure the following fields for each claim:
```

	- **Name**: OIDC claim name
	- **Scope**: Select the OIDC scope where this claim should be included. In most cases, we recommend selecting `profile` since it already includes other custom claims from the IdP.
	- **IdP claim**: The identity provider value that should map to this OIDC claim. You can select any [SAML attribute](/cloudflare-one/identity/idp-integration/generic-saml/#saml-headers-and-attributes) or [OIDC claim](/cloudflare-one/identity/idp-integration/generic-oidc/#oidc-claims) that was configured in a Zero Trust IdP integration.
	- **Required**: If a claim is marked as required but is not provided by an IdP, Cloudflare will fail the authentication request and show an error page.
	- **Add per IdP claim**: (Optional) If you turned on multiple identity providers for the SaaS application, you can choose different attribute mappings for each IdP. These values will override the parent **IdP claim**.

## Advanced settings

### Access token lifetime

> The OIDC Access token authorizes users to connect to the SaaS application through Cloudflare Access. You can set an **Access token lifetime** to determine how often Cloudflare should authenticate the user to your SaaS application. To balance security and user convenience, you can configure a short token lifetime in conjunction with a longer **Refresh token lifetime**. When the access token expires, Cloudflare will use the refresh token to obtain a new access token after checking the user's identity against your Access policies. When the refresh token expires, the user will need to log back in to the identity provider. The refresh token lifetime should be less than your [global session duration](/cloudflare-one/identity/users/session-management/), otherwise the global session would take precedence.

Reviewer: ```suggestion
The OIDC Access token authorizes users to connect to the SaaS application through Cloudflare Access. You can set an **Access token lifetime** to determine the window in which the token can be used to establish authentication with the SaaS application —if it expires, the user must reauthenticate through Cloudflare Access. To balance security and user convenience, Cloudflare recommends you configure a short Access token lifetime in conjunction with a longer **Refresh token lifetime** if supported by your application. When the access token expires, Cloudflare will use the refresh token to obtain a new access token after checking the user's identity against your Access policies. When the refresh token expires, the user will need to log back in to the identity provider. The refresh token lifetime should be less than your [global session duration](/cloudflare-one/identity/users/session-management/), otherwise the global session would take precedence.
```

:::note
<Render file="access/saas-apps/saas-sessions" params={{ session: "OIDC Access tokens", token: "Access token"}} />
:::

### OIDC flows

Some SaaS applications require SSO providers to provide tokens to the browser without backend authentication. Access for SaaS supports the following OIDC flows:

- **No additional OIDC flows**: (Default) Recommended unless your application requires additional flows.
- **Hybrid flows**: Used by applications that require information from the ID token before authenticating the user.
- **Implicit flows**: (Not recommended) Typically used by frontend applications that cannot store secrets and which do not support **PKCE without client secret**.

> Cloudflare may include different `response_type` values in the authorization request depending on the selected flow. For example, the implicit flow allows Cloudflare to send the ID token, Access token, or both the ID token and Access token.

Reviewer: ```suggestion
Cloudflare allows various `response_type` values in the authorization request depending on the selected flow. For example, the implicit flow allows Cloudflare to return the ID token, Access token, or both the ID token and Access token from the Authorization endpoint.
```
![Screenshot 2025-03-10 at 4 34 28 PM](https://github.com/user-attachments/assets/0cbe4d3c-7c8e-440d-8440-65d34d8234fc)
- just noting these two toggles as well. Hybrid and Implicit both have these two toggles. They select whether we allow returning id_token and/or token from the Authorization endpoint.

- add a note that refresh tokens are not supported with Hybrid or Implicit flow.

## OIDC claims

OIDC claims refer to the user identity characteristics that Cloudflare Access shares with your OIDC SaaS application upon successful authentication. An OIDC scope defines a set of OIDC claims. By default, Cloudflare Access passes all [standard claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) that are included in the `openid`, `email`, `profile`, and `groups` scopes (if available).

   | Scope     | Description                                          |
   | --------- | ---------------------------------------------------- |
   | `openid`  | Includes a unique identifier for the user (required). |
   | `email`   | Includes the user's email address.                    |
   | `profile` | Includes the user's name and all custom OIDC claims from the IdP.         |
   | `groups`  | Include the user's IdP group membership.             |

In your Access application, you can configure the OIDC scopes and claims that Access sends to the SaaS provider. For example, you can remove the `groups` scope if your SaaS application does not need to receive user group information.

### Filter groups

> In **Group filter regex**, you can enter a regular expression to define the identity provider groups that you want to include in the `groups` scope. For example, if you enter the expression `(^TEAM-Engineering-.$)|(^TEAM-Product-.$)`, only groups with names like  TEAM-Engineering-A or TEAM-Product-B would get passed to the SaaS application.

Reviewer 2: ```suggestion
In **Group filter regex**, you can enter a regular expression to define the identity provider groups that you want to include in the `groups` scope. For example, if you enter the expression `(^TEAM-Engineering-.$)|(^TEAM-Product-.$)`, only groups with names like TEAM-Engineering-A or TEAM-Product-B would get passed to the SaaS application.
```
